riscv/src/elf.rs

```diff
@@ -0,0 +1,1229 @@
+                    }
+                }
+            }
+            _ => return None,
+        };
+
+        // TODO: implement here other kinds of RISC-V fusions as optimization.
+
+        if let (ReadOrWrite::Write(refs), HighLevelImmediate::CodeLabel(addr)) =
+            (&mut self.referenced_text_addrs, &result.args.imm)
+        {
+            refs.insert(*addr);
+        }
+
+        Some(result)
+    }
+
+    fn map_one(&mut self, insn: MaybeInstruction) -> HighLevelInsn {
+        let original_address = insn.address;
+        let Some(insn) = insn.insn else {
+            return HighLevelInsn {
+                op: "unimp",
+                args: Default::default(),
+                original_address,
+            };
+        };
+
+        let mut imm = match insn.opc {
+            // All jump instructions that have the immediate as an address
+            Op::JAL | Op::BEQ | Op::BNE | Op::BLT | Op::BGE | Op::BLTU | Op::BGEU => {
+                let addr = (insn.imm.unwrap() + original_address as i32) as u32;
+                if let ReadOrWrite::Write(refs) = &mut self.referenced_text_addrs {
+                    refs.insert(addr);
+                }
+
+                HighLevelImmediate::CodeLabel(addr)
+            }
+            // We currently only support standalone jalr if offset is zero
+            Op::JALR => {
+                assert!(
+                    insn.imm.unwrap() == 0,
+                    "jalr with non-zero offset is not supported"
+                );
+
+                HighLevelImmediate::Value(0)
+            }
+            // LUI is special because the decoder already shifts the immediate,
+            // but the code gen expects it unshifted, so we have to undo.
+            Op::LUI => HighLevelImmediate::Value(insn.imm.unwrap() >> 12),
+            // We don't support arbitrary AUIPCs, but it is trivial to transform
+            // one to an LI. If it passed the two-by-two transformation and got
+            // here, this is a reference to data, so it is safe to transform it.
+            Op::AUIPC => {
+                return HighLevelInsn {
+                    op: "li",
+                    args: HighLevelArgs {
+                        rd: insn.rd.map(|x| x as u32),
+                        imm: HighLevelImmediate::Value(
+                            insn.imm.unwrap().wrapping_add(original_address as i32),
+                        ),
+                        ..Default::default()
+                    },
+                    original_address,
+                };
+            }
+            // All other instructions, which have the immediate as a value
+            _ => match insn.imm {
+                Some(imm) => HighLevelImmediate::Value(imm),
+                None => HighLevelImmediate::None,
+            },
+        };
+
+        // For some reason, atomic instructions comes with the immediate set to
+        // zero instead of None (maybe to mimic assembly syntax? Who knows). Fix
+        // this.
```

Comment:
fix in this PR?

Reply:
The fix is the code following. I reworded.